librad/src/git/types.rs

```diff
@@ -44,15 +45,47 @@ impl Display for RefsCategory {
     }
 }
 
+/// Type witness for a [`Reference`] that should point to a single reference.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Single;
+
+/// Type witness for a [`Reference`] that should point to multiple references.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Multiple;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum SomeReference {
+    Single(Reference<Single>),
+    Multiple(Reference<Multiple>),
+}
+
+impl Display for SomeReference {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Single(reference) => write!(f, "{}", reference),
+            Self::Multiple(reference) => write!(f, "{}", reference),
+        }
+    }
+}
+
+/// A structure for building a git reference.
 #[derive(Clone, Debug, PartialEq)]
-pub struct Reference {
+pub struct Reference<N> {
+    /// The namespace the reference lives under.
     pub namespace: Namespace,
+    /// The remote peer the reference might live under.
     pub remote: Option<PeerId>,
+    /// The category of the reference.
     pub category: RefsCategory,
+    /// The suffix path of the reference, e.g. `self`, `id`, `banana/pineapple`.
     pub name: String, // TODO: apply validation like `uri::Path`
+
+    /// Type witness for the cardinality this reference could point to, i.e. if
+    /// we are pointing to exactly one, zero or more, or both.
+    marker: PhantomData<N>,
```

Comment:
Interestingly, I remember clippy did want me to prefix such fields with an underscore... maybe because it thinks the method name `new` is special, and if the field is only used there, it must be unused?

Reply:
Clippy :shrug: